go/dpagg/mean_test.go

```diff
@@ -738,3 +738,153 @@ func TestBMFloat64Serialization(t *testing.T) {
+		},
+		// Clamp too low bounds.
+		{
+			meanOpt:      &BoundedMeanFloat64Options{Lower: 1, Upper: 9, Noise: noNoise, MaxContributionsPerPartition: 1},
+			sumConfInt:   noise.ConfidenceInterval{LowerBound: -100, UpperBound: -50},
+			countConfInt: noise.ConfidenceInterval{LowerBound: 2, UpperBound: 5},
+			want:         noise.ConfidenceInterval{LowerBound: 1, UpperBound: 1},
+		},
+		// Clamp too high bounds.
+		{
+			meanOpt:      &BoundedMeanFloat64Options{Lower: 1, Upper: 9, Noise: noNoise, MaxContributionsPerPartition: 1},
+			sumConfInt:   noise.ConfidenceInterval{LowerBound: 70, UpperBound: 100},
+			countConfInt: noise.ConfidenceInterval{LowerBound: 2, UpperBound: 5},
+			want:         noise.ConfidenceInterval{LowerBound: 9, UpperBound: 9},
+		},
+	} {
+		mean := NewBoundedMeanFloat64(tc.meanOpt)
+		mean.normalizedSum.noise = getMockConfInt(tc.sumConfInt)
+		mean.count.noise = getMockConfInt(tc.countConfInt)
+		mean.Result()
+		got, _ := mean.computeConfidenceIntervalSplit(0.1, 0.05) //parameters are ignored
+		if !ApproxEqual(got.LowerBound, tc.want.LowerBound) {
+			t.Errorf("TestMeanComputeConfidenceIntervalSplit(ConfidenceInterval{%f, %f})=%0.10f, want %0.10f, LowerBounds are not equal",
+				tc.meanOpt.Lower, tc.meanOpt.Upper, got.LowerBound, tc.want.LowerBound)
+		}
+		if !ApproxEqual(got.UpperBound, tc.want.UpperBound) {
+			t.Errorf("TestMeanComputeConfidenceIntervalSplit(ConfidenceInterval{%f, %f})=%0.10f, want %0.10f, UpperBounds are not equal",
+				tc.meanOpt.Lower, tc.meanOpt.Upper, got.UpperBound, tc.want.UpperBound)
+		}
+	}
+}
+
+/**
+ * This test was designed to be not deterministic. It goes along with deterministic analogues in
+ * order to ensure that they don't miss something.
+ */
+func TestMeanComputeConfidenceInterval_confidenceBoundsAlwaysInsideProvidedBoundaries(t *testing.T) {
+	{
+		datasize := 10
+		for i := 0; i < 100; i++ {
+			lower := rand.Uniform() * 100
+			upper := lower + rand.Uniform()*100
+			meanOpt := &BoundedMeanFloat64Options{Epsilon: 1.0, Delta: 0.123, Lower: lower, Upper: upper, Noise: noise.Gaussian(), MaxContributionsPerPartition: 1}
+			mean := NewBoundedMeanFloat64(meanOpt)
+			for j := 0; j < datasize; j++ {
+				mean.Add(rand.Uniform() * 300 * rand.Sign())
+			}
+			mean.Result()
+			// Generate a random meanAlpha and a random split.
+			meanAlpha := rand.Uniform()
+			countAlpha := rand.Uniform() * meanAlpha
+			meanConfInt, _ := mean.computeConfidenceIntervalSplit(meanAlpha, countAlpha)
+			if meanConfInt.LowerBound < lower || meanConfInt.LowerBound > upper {
+				t.Errorf("TestMeanComputeConfidenceInterval_confidenceBoundsAlwaysInsideProvidedBoundaries")
+			}
+			if meanConfInt.UpperBound < lower || meanConfInt.UpperBound > upper {
+				t.Errorf("TestMeanComputeConfidenceInterval_confidenceBoundsAlwaysInsideProvidedBoundaries")
+			}
+
+		}
+	}
+}
+
+/**
+ * This test was designed to be not deterministic. Brute force for optimal ratio should always
```

Comment:
A general guideline. Don't do random unit tests just becuase you can't think of plausible test parameters or because you think you will get more test coverage.

The only reason to have randomness in your unit test is if you want to test a statistical property. And in that case you should have clear bounds on the success/failure probability of the test

Reply:
I understand, therefore would it be best to set some specific bounds and add specific numbers like the previous test? (only one deterministic test)

Reply:
Yes, specific bounds and parametrs would be better 👍